crates/in_stream/src/trait.rs

```diff
@@ -0,0 +1,143 @@
+    /// given an any, attempt to downcast to a concrete type
+    fn from_any(any: InStreamConfigAny) -> Result<Self> {
+        match any.downcast::<Self>() {
+            Ok(v) => Ok(*v),
+            Err(_) => Err(Error::new(ErrorKind::InvalidInput, "bad config type")),
+        }
+    }
+
+    /// convert a generic type into a concrete one
+    fn from_gen<C: InStreamConfig>(c: C) -> Result<Self> {
+        Self::from_any(c.to_any())
+    }
+
+    /// convert this type into an any
+    fn to_any(self) -> InStreamConfigAny {
+        Box::new(self)
+    }
+}
+
+// allow unit type to be used for configuration
+impl InStreamConfig for () {}
+
+// if you've already got an any, we also want to be able to use that
+impl InStreamConfig for InStreamConfigAny {
+    fn to_any(self) -> InStreamConfigAny {
+        // we're already a box, don't re-box
+        self
+    }
+}
+
+/// implement this trait to provide listening/server socket type functionality
+pub trait InStreamListener<R: Sized + Debug + Send + Sync, W: Sized + Debug + Send + Sync>:
+    Sized + Debug + Send + Sync
+{
+    type Stream: InStream<R, W>;
+
+    /// begin listening on the given url spec
+    /// this function does the actual work of binding, but it is recommended
+    /// your struct provide a wrapper with a concrete config type
+    fn raw_bind<C: InStreamConfig>(url: &Url2, config: C) -> Result<Self>;
+
+    /// access the url for the bound interface
+    fn binding(&self) -> Url2;
+
+    /// attempt to accept a stream/socket from this binding
+    /// may return Err(ErrorKind::WouldBlock.into())
+    fn accept(&mut self) -> Result<<Self as InStreamListener<R, W>>::Stream>;
+}
+
+/// implement this if your listener accepts std::io::{Read, Write} streams
+pub trait InStreamListenerStd
+where
+    for<'a> Self: InStreamListener<&'a mut [u8], &'a [u8]>,
+{
+    type StreamStd: InStreamStd;
+
+    /// use `accept_std` if you want your streams to work with byte data
+    fn accept_std(&mut self) -> Result<<Self as InStreamListenerStd>::StreamStd>;
+}
+
+/// implement this trait to provide a stream endpoint / socket type connection
+/// works like combined std::io::{Read, Write}, but with generic types
+/// the underlying stream should be treated as non-blocking.
+/// For example, if this is a TLS stream, we may still need to complete
+/// a handshaking process before data can actually be written.
+///
+/// `read` will return WouldBlock if there is no pending data to be read
+/// `write` will buffer any data if the stream is not ready to write
+/// `flush` will block until pending data has been written
+pub trait InStream<R: Sized + Debug + Send + Sync, W: Sized + Debug + Send + Sync>:
+    Sized + Debug + Send + Sync
+{
+    /// your implementation should work with a single url scheme/protocol
+    /// if you need to support multiple (i.e. ws:// vs wss://) consider
+    /// using macros/code generation to DRY
+    const URL_SCHEME: &'static str;
+
```

Comment:
Traits can define constants? Good to know!

Reply:
:v: